osbs/utils.py

```diff
@@ -190,72 +192,120 @@ def buildconfig_update(orig, new, remove_nonexistent_keys=False):
-    for the cloned repository.
-
-    :param git_uri: string, git repository URI
-    :param git_ref: string, target git reference
-    :param git_branch: string, target git branch
-    :param max_retries: int, maximum number of git clone retries
-    :param backoff_factor: int, exponential backoff factor for retries, in seconds
+
+def reset_git_repo(target_dir, git_reference, retry_depth=None):
     """
-    tmpdir = tempfile.mkdtemp()
-    repo_path = os.path.join(tmpdir, "repo")
-    # when you clone into an empty directory and cloning process fails
-    # git will remove the empty directory (why?!)
-    args = ['git', 'clone', git_uri]
-    if git_branch:
-        args += ['-b', git_branch]
-
-    args.append(repo_path)
-    try:
-        try:
-            subprocess_check_output_with_retry(args, max_retries, backoff_factor)
-        except subprocess.CalledProcessError as ex:
-            raise OsbsException("Unable to clone git repo '%s' "
-                                "branch '%s'" % (git_uri, git_branch),
-                                cause=ex, traceback=sys.exc_info()[2])
+    hard reset git clone in target_dir to given git_reference
 
-        # Find the specific ref we want
+    :param target_dir: str, filesystem path where the repo is cloned
+    :param git_reference: str, any valid git reference
+    :param retry_deeper: int, if the repo was cloned with --shallow, this is the expected
+                         depth of the commit
+    :return: str, commit ID of HEAD
+    """
+    deepen = retry_depth or 0
+    commit_depth = None
+    for _ in range(1, 10):
         try:
-            subprocess.check_output(['git', 'reset', '--hard'], cwd=repo_path)
-            subprocess.check_output(['git', 'checkout', git_ref], cwd=repo_path)
-        except subprocess.CalledProcessError as ex:
-            raise OsbsException("Unable to reset branch to '%s'" % git_ref,
-                                cause=ex, traceback=sys.exc_info()[2])
+            if not deepen:
+                cmd = ['git', 'rev-list', '--count', '{0}'.format(git_reference)]
+                commit_depth = int(subprocess.check_output(cmd, cwd=target_dir))
+            cmd = ["git", "reset", "--hard", git_reference]
+            logger.debug("Resetting current HEAD: '%s'", cmd)
+            subprocess.check_call(cmd, cwd=target_dir)
+            break
+        except subprocess.CalledProcessError:
+            if not deepen:
+                raise OsbsException('cannot find commit %s in repo %s' %
+                                    (git_reference, target_dir))
+            deepen *= 2
+            cmd = ["git", "fetch", "--depth", "{0}".format(deepen)]
+            subprocess.check_call(cmd, cwd=target_dir)
+    else:
+        raise OsbsException('cannot find commit %s in repo %s' % (git_reference, target_dir))
 
-        yield repo_path
+    cmd = ["git", "rev-parse", "HEAD"]
+    logger.debug("getting SHA-1 of provided ref '%s'", cmd)
```

Comment:
Looks like this will log `getting SHA-1 of provided ref '['git', 'rev-parse', 'HEAD']'`. Is that useful? Also, the command does not retrieve the hash for the *provided ref*, but for the commit we reset to... which the caller already knows.

I'm not sure why we're returning this value (or running the command).

Reply:
`reset_git_repo()` can take any valid git_reference, so things like HEAD^ can be the argument, in which case you don't know the hash.

I'm not sure I follow your objection.  We reset the HEAD to some arbitrary git reference, which may or may not be a hash, and then we return the hash of the new head.
ie:
```
$ git reset --hard master
HEAD is now at 031467b Merge pull request #112 from rcerven/metadata_for_scratch
$ git rev-parse HEAD
031467b48f8139935a7d4ab8b6ded9c30b653529
$ git reset --hard HEAD^^^
HEAD is now at 199d96c Automatic commit of package [koji-containerbuild] minor release [0.7.12-1].
$ git rev-parse HEAD
199d96ca229edfcf9f81fca4c5304823bfebd2f6
```

Reply:
Ah, git_reference is not necessary a hash -- I had overlooked that. Thanks.
But is `getting SHA-1 of provided ref '['git', 'rev-parse', 'HEAD']'` what we want to log? Do we want to log what the provided ref actually was?